[WIP]ConsoleInteraction: Show context for lines

Sanny26 opened the pull request:
When results are displayed on console, diff and lines are printed.
To make it easier and understandable, this commit adds changes to
print context of these lines to console. (Completed only one part
of issue)

Closes https://github.com/coala/coala/issues/2745

<!--
Thanks for your contribution!

Reviewing pull requests takes a lot of time and we're all volunteers. Please make sure you go through the following checklist and all items before pinging someone for a review.
-->

### Checklist

- [ ] I have rebased properly. Please see [our tutorial on rebasing](http://coala.io/git#rebasing).
- [ ] I have gone through the [commit guidelines](http://coala.io/commit) and I've followed them.
- [ ] I have followed the [guidelines for the commit shortlog](http://coala.io/commit#shortlog).
- [ ] I have followed the [guidelines for the commit body](http://coala.io/commit#commit-body).
- [ ] I have included the issue URL with the 'Fixes' keyword if the commit fixes a *real bug* and the 'Closes' keyword if it adds a feature or enhancement. For details, check the [issue reference section in our commit guidelines](http://coala.io/commit#issue-reference).
- [ ] I have [run all the tests](http://api.coala.io/en/latest/Developers/Executing_Tests.html) and they all pass. If any CI (below) is not green, please fix them. If GitMate issues appear you will have to amend your commits, pushing new commits on top is not sufficient!

**After you submit your pull request, DO NOT click the 'Update Branch' button**

### Reviewers

<!--
Please list the Github handles of people you think susceptible to review this PR. Feel free to leave this section blank if you don't know who to tag.
-->

<!--
End note:

As you learn things over your Pull Request please help others on the chat and on PRs to get their stuff right as well!
-->

sims1253 commented:
could you post a screenshot of how it looks with the changes?

Sanny26 commented:
@sims1253 I have been using unit tests only to see how my output matches the expected output. How can I test these changes?

sims1253 commented:
run your version of coala over the test file for example

Sanny26 commented:
@sims1253  Screenshot link: http://imgur.com/a/tnOot
Used code snippet link: http://pastebin.com/ZXGpp7cU
Anything that I should change?

Udayan12167 commented:
Remove from the commit shortlog

Udayan12167 commented:
Don't use closes in the PR if doesn't completely solve the issue. Use `See <issue-number>`

Sanny26 commented:
@Udayan12167  removed it :)
Circleci fails, showing that coverage is not 100% for my additions and the BrPart is marked 1, which is for branch coverage right? So, can I add '#pragma : no branch' for the for loop? Or should I make any other change to get 100% coverage?

Udayan12167 commented:
You should get coverage to 100%.

Sanny26 commented:
Results View: http://imgur.com/a/a36zQ
Does everything look fine? Should I continue to write tests for the changes?

sils commented on coalib/results/result_actions/ResultAction.py:
```diff
@@ -61,7 +62,16 @@ def apply_from_section(self,
+        else:
+            return self.apply(result,
+                              original_file_dict,
+                              file_diff_dict, **params)
```
I don't think adding this here makes sense, this information is only valid for a given diff and should be generated on application of the showpatchaction or so, this has nothing to do in the class scope

sils commented:
@Sanny26 hey this is a good PR, thanks for pinging me about this. Please read :point_up: [February 15, 2017 5:34 PM](https://gitter.im/coala/coala?at=58a483191465c46a566bb9f5) and try to simplify your code. I'm very interested in this change, you may mention me on gitter if you need a review for this when the gitmate issues are fixed and the code is simplified!

Sanny26 commented:
@sils As you mentioned on gitter, I'm searching for the required lines and their line numbers. But I'm searching for them, just before printing the affected line and storing in a list. I'm returning this list and passing it wherever it is needed. I'm using the sourcerange of the affected line in print_lines() (https://github.com/coala/coala/pull/3659/files#diff-5be6f2c646fedbab1fd2aea142f54303R172) to search.
This chunk can be moved to a new function, but I'll still have to return the list from the print_lines(). So, the change would be to move it to new function, and the rest of it remains same.

Alternative that I could think of: Call a function in print_results() before calling affected_files()(https://github.com/coala/coala/pull/3659/files#diff-5be6f2c646fedbab1fd2aea142f54303R485), but then there are some warnings/error messages that can be raised in affected_files(), would it be correct to call this context function before these warnings checks are passed?
So, what do you suggest? Is any of these the right place to add the changes?

Sanny26 commented:
@sils I have made a few changes:
- Return error in print_affected_files() is corrected. But I could not get any idea on what files coala can be run to know if the changes work fine. Could you suggest one?
https://github.com/coala/coala/pull/3659/files#diff-5be6f2c646fedbab1fd2aea142f54303R455
- Added apply() and get_metadata() to ShowPatchAction . Is this the change that you asked for?
- Removed '@' and used '|' for printing, as was suggested
- Created get_context()  (https://github.com/coala/coala/pull/3659/files#diff-5be6f2c646fedbab1fd2aea142f54303R170) 
In this function, if there is no context, i.e., the line has zero indentation, then I added the prev 2 lines as context. And also, if line has trailing tab space, for comparison sake I replaced it with spaces (1 tab=8 space). Otherwise, context might not be as expected or it could be unclean
Any suggestions on these changes?
